<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Igor Ruiz de los Mozos">

<title>R Programming Basics ‚Äì Hands-On Practical</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="R_Programming_basics_practical_files/libs/clipboard/clipboard.min.js"></script>
<script src="R_Programming_basics_practical_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="R_Programming_basics_practical_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="R_Programming_basics_practical_files/libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="R_Programming_basics_practical_files/libs/quarto-html/popper.min.js"></script>
<script src="R_Programming_basics_practical_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="R_Programming_basics_practical_files/libs/quarto-html/anchor.min.js"></script>
<link href="R_Programming_basics_practical_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="R_Programming_basics_practical_files/libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="R_Programming_basics_practical_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="R_Programming_basics_practical_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="R_Programming_basics_practical_files/libs/bootstrap/bootstrap-298968afe97ad68a9ad451e6e9bc2c85.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-full">

<main class="content column-page" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">R Programming Basics ‚Äì Hands-On Practical</h1>
</div>



<div class="quarto-title-meta column-page">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Igor Ruiz de los Mozos </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<blockquote class="blockquote">
<p><strong>Programme:</strong> Master‚Äôs in Biomedical Engineering‚ÄÉ‚Ä¢‚ÄÉ<strong>Module:</strong> Advanced Bioinformatics<br>
<strong>Lesson </strong> R Fundamentals, Data Wrangling &amp; Visualisation</p>
</blockquote>
<hr>
<section id="table-of-contents" class="level2">
<h2 class="anchored" data-anchor-id="table-of-contents">üó∫Ô∏è Table of Contents</h2>
<ol type="1">
<li><a href="#overview">Overview &amp; Learning Objectives</a><br>
</li>
<li><a href="#rstudio">The RStudio Interface</a><br>
</li>
<li><a href="#basics">R Programming Basics</a><br>
</li>
<li><a href="#structures">Exploring Objects &amp; Structures</a><br>
</li>
<li><a href="#dplyr">Wrangling with <strong>dplyr</strong></a><br>
</li>
<li><a href="#tidyr">Reshaping with <strong>tidyr</strong></a><br>
</li>
<li><a href="#ggplot">Visualising with <strong>ggplot2</strong></a><br>
</li>
<li><a href="#workflow">üåü Full Workflow on <code>biomed_data.csv</code></a></li>
</ol>
<hr>
</section>
<section id="overview-learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="overview-learning-objectives">1 <span id="overview">Overview &amp; Learning Objectives</span></h2>
<section id="introduction" class="level3">
<h3 class="anchored" data-anchor-id="introduction">Introduction</h3>
<p>This practical introduces the foundational concepts of R programming and its application in bioinformatics. R is a powerful language for statistical computing and data visualization, widely used in biomedical research.</p>
</section>
<section id="goals" class="level3">
<h3 class="anchored" data-anchor-id="goals">Goals</h3>
<p>By the end of this session, you will be able to: - Navigate the RStudio interface and understand its key components. - Write and execute basic R code for data manipulation and visualization. - Use <strong>dplyr</strong> verbs to filter, summarize, and join datasets. - Reshape data using <strong>tidyr</strong> for better analysis. - Create a variety of plots using <strong>ggplot2</strong>, including histograms, scatter plots, and heatmaps. - Apply these skills to analyze a realistic dataset (<code>biomed_data.csv</code>) containing information on 200 patients.</p>
</section>
<section id="why-learn-r" class="level3">
<h3 class="anchored" data-anchor-id="why-learn-r">Why Learn R?</h3>
<p>R is an essential tool for bioinformatics due to its flexibility, extensive libraries, and active community. It allows researchers to handle large datasets, perform complex statistical analyses, and create publication-quality visualizations.</p>
<hr>
</section>
</section>
<section id="the-rstudio-interface" class="level2">
<h2 class="anchored" data-anchor-id="the-rstudio-interface">2 <span id="rstudio">The RStudio Interface</span></h2>
<table class="caption-top table">
<colgroup>
<col style="width: 35%">
<col style="width: 31%">
<col style="width: 33%">
</colgroup>
<thead>
<tr class="header">
<th>Pane</th>
<th>Purpose</th>
<th>Tip</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Source/Editor</strong> (top-left)</td>
<td>Write &amp; save scripts (<code>.R</code>, <code>.Rmd</code>)</td>
<td><code>Ctrl/Cmd + Shift + C</code> toggles comments</td>
</tr>
<tr class="even">
<td><strong>Console</strong> (bottom-left)</td>
<td>Run commands interactively</td>
<td><code>‚Üë/‚Üì</code> scrolls through history</td>
</tr>
<tr class="odd">
<td><strong>Environment / History</strong> (top-right)</td>
<td>View objects &amp; past commands</td>
<td>Click ‚úñ to remove objects</td>
</tr>
<tr class="even">
<td><strong>Files / Plots / Packages / Help</strong></td>
<td>File nav, graphics, install pkgs, docs</td>
<td>Click a plot ‚Üí <em>Export</em> to save</td>
</tr>
</tbody>
</table>
<blockquote class="blockquote">
<p><strong>Tip:</strong> Use the shortcut <code>Ctrl + Enter</code> (Windows) or <code>Cmd + Enter</code> (Mac) to run a line of code from the Source Pane in the Console.</p>
</blockquote>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span>opts_knit<span class="sc">$</span><span class="fu">set</span>(<span class="at">root.dir =</span> <span class="st">"/workspaces/codespaces_RStudio/course_materials/R_basics"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Set your working directory once per session</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/workspaces/codespaces_RStudio/practicals"</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"/workspaces/codespaces_RStudio/course_materials/R_basics"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Install required libraries once:</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="fu">c</span>(<span class="st">"tidyverse"</span>, <span class="st">"viridis"</span>, <span class="st">"GGally"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Installing packages into '/usr/local/lib/R/site-library'
(as 'lib' is unspecified)</code></pre>
</div>
</div>
</section>
<section id="r-programming-basics" class="level2">
<h2 class="anchored" data-anchor-id="r-programming-basics">3 <span id="basics">R Programming Basics</span></h2>
<section id="writing-your-first-r-code" class="level3">
<h3 class="anchored" data-anchor-id="writing-your-first-r-code">Writing Your First R Code</h3>
<p>Let‚Äôs start with some simple R commands. Open the R Console and try the following:</p>
</section>
<section id="console-vs-editor" class="level3">
<h3 class="anchored" data-anchor-id="console-vs-editor">3.1 Console vs Editor</h3>
<ul>
<li><p>Console for quick commands.</p></li>
<li><p>Editor for scripts.</p></li>
<li><p>Execute lines with Ctrl/Cmd + Enter.</p></li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Basic arithmetic</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> <span class="sc">+</span> <span class="dv">2</span>  <span class="co"># Addition</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span> <span class="sc">*</span> <span class="dv">3</span>  <span class="co"># Multiplication</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 15</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> <span class="sc">/</span> <span class="dv">2</span> <span class="co"># Division</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5</code></pre>
</div>
</div>
<p>Explanation Comments: Lines starting with <code>#</code> are comments and are ignored by R. Assignment: The <code>&lt;-</code> operator assigns values to variables. You can also use <code>=</code> but <code>&lt;-</code> is preferred in R. Execution: Type the code in the Console or write it in a script and run it. Exercise: Try assigning different values to x and y and observe how z changes.</p>
</section>
<section id="variable-assignment-arithmetic" class="level3">
<h3 class="anchored" data-anchor-id="variable-assignment-arithmetic">3.2 Variable Assignment &amp; Arithmetic</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">10</span>    <span class="co"># arrow assignment </span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>y <span class="ot">=</span> <span class="dv">5</span>      <span class="co"># equals also valid</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>x</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>x <span class="sc">+</span> y      <span class="co"># addition</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 15</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>x <span class="sc">*</span> y      <span class="co"># multiplication</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 50</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>x <span class="sc">^</span> <span class="dv">2</span>      <span class="co"># exponentiation</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 100</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>x <span class="sc">&gt;</span> y      <span class="co"># logical test</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>Explanation Comments: Lines starting with <code>#</code> are comments and are ignored by R. Assignment: The <code>&lt;-</code> operator assigns values to variables. You can also use = but &lt;- is preferred in R for clarity and consistency. Arithmetic Operations: R supports basic arithmetic operations like addition (<code>+</code>), multiplication (<code>*</code>), and exponentiation (<code>^</code>). Logical Tests: Logical operators like &gt; return TRUE or FALSE based on the comparison. Exercise: Assign different values to x and y and observe how the results of x + y and x * y change. Try other logical operators like <code>&lt;</code>, <code>==</code>, and <code>!=</code> to compare x and y.</p>
</section>
<section id="data-types-coercion" class="level3">
<h3 class="anchored" data-anchor-id="data-types-coercion">3.3 Data Types &amp; Coercion</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>my_numeric <span class="ot">&lt;-</span> <span class="fl">42.0</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>my_char    <span class="ot">&lt;-</span> <span class="st">"Hello"</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>my_logical <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(my_numeric)   <span class="co"># "numeric"</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numeric"</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>my_numeric <span class="ot">&lt;-</span> <span class="fu">as.character</span>(my_numeric)  <span class="co"># "42"</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Explanation Data Types: R has several basic data types, including: numeric: Numbers with or without decimals. character: Text or strings. logical: Boolean values (TRUE or FALSE). Type Coercion: Functions like as.character() can convert one data type to another. For example, a numeric value can be converted to a character string. Exercise: Create variables of different data types (e.g., integer, factor) and use the class() function to check their types. Try coercing a character string to numeric using as.numeric() and observe the result.</p>
</section>
<section id="exploring-objects" class="level3">
<h3 class="anchored" data-anchor-id="exploring-objects">3.4 Exploring Objects</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>v <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(v)         <span class="co"># structure</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> num [1:5] 1 2 3 4 5</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(v)     <span class="co"># summary stats</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
      1       2       3       3       4       5 </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(v)      <span class="co"># 5</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5</code></pre>
</div>
</div>
<p>Explanation Vectors: The <code>c()</code> function combines values into a vector. Vectors are one of the most basic data structures in R. Functions: R provides built-in functions like <code>length()</code>, <code>sum()</code>, and <code>mean()</code> to operate on vectors. Exercise: Create a vector with at least 10 elements and calculate its sum, mean, and length. Try creating a character vector (e.g., c(‚ÄúA‚Äù, ‚ÄúB‚Äù, ‚ÄúC‚Äù)) and observe what happens when you use sum() or mean() on it.</p>
</section>
</section>
<section id="exploring-objects-structures" class="level2">
<h2 class="anchored" data-anchor-id="exploring-objects-structures">4 <span id="structures">Exploring Objects &amp; Structures</span></h2>
<section id="vectors" class="level4">
<h4 class="anchored" data-anchor-id="vectors">Vectors</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>v <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">20</span>, <span class="dv">30</span>, <span class="dv">40</span>, <span class="dv">50</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>v[<span class="dv">1</span>]       <span class="co"># Access the first element</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>v[<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>]     <span class="co"># Access elements from index 2 to 4</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 20 30 40</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>v[<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">5</span>)] <span class="co"># Access the 1st and 5th elements</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10 50</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>v[v <span class="sc">&gt;</span> <span class="dv">25</span>]  <span class="co"># Subset elements greater than 25</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 30 40 50</code></pre>
</div>
</div>
<p>Explanation Indexing: Use square brackets <code>[]</code> to access specific elements of a vector. Indexing in R starts at 1 (not 0). Subsetting: You can subset vectors using logical conditions (e.g., v &gt; 25). Exercise: Create a vector of your choice and practice accessing individual elements and ranges. Use logical conditions to subset elements based on specific criteria (e.g., values less than 15).</p>
</section>
<section id="matrices" class="level4">
<h4 class="anchored" data-anchor-id="matrices">Matrices</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a 3x3 matrix</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>, <span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>m[<span class="dv">2</span>, <span class="dv">3</span>]         <span class="co"># Access the element in the 2nd row, 3rd column</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(m)          <span class="co"># Get the dimensions of the matrix (rows and columns)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3 3</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>m[<span class="dv">1</span>, ]          <span class="co"># Access the entire first row</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1 2 3</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>m[, <span class="dv">2</span>]          <span class="co"># Access the entire second column</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2 5 8</code></pre>
</div>
</div>
<p>Explanation Matrix Creation: The <code>matrix()</code> function creates a matrix. The nrow argument specifies the number of rows, and byrow = TRUE fills the matrix row-wise. Indexing: Use <code>[row, column]</code> to access specific elements. Leaving the row or column blank (e.g., <code>[1, ]</code>) selects all elements in that dimension. Dimensions: The <code>dim()</code> function returns the dimensions of the matrix as a vector (number of rows and columns). Exercise: Create a 4x4 matrix with numbers from 1 to 16. Access the element in the 3rd row and 4th column. Extract the 2nd row and the 3rd column separately. Try adding two matrices of the same dimensions.</p>
</section>
<section id="lists" class="level4">
<h4 class="anchored" data-anchor-id="lists">Lists</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a list containing a vector and a matrix</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>my_list <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">a =</span> v, <span class="at">b =</span> m)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>my_list<span class="sc">$</span>a      <span class="co"># Access the vector by name</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10 20 30 40 50</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>my_list<span class="sc">$</span>b      <span class="co"># Access the matrix by name</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] [,3]
[1,]    1    2    3
[2,]    4    5    6
[3,]    7    8    9</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>my_list[[<span class="dv">1</span>]]   <span class="co"># Access the first element of the list (vector)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10 20 30 40 50</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>my_list[[<span class="dv">2</span>]][<span class="dv">2</span>, <span class="dv">3</span>]  <span class="co"># Access the 2nd row, 3rd column of the matrix in the list</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6</code></pre>
</div>
</div>
<p>Explanation Lists: A list is a flexible data structure that can contain elements of different types (e.g., vectors, matrices, data frames). Accessing Elements: Use <code>$</code> to access elements by name or <code>[[ ]]</code> to access elements by position. You can also combine indexing to access specific parts of nested elements. Exercise: Create a list containing a numeric vector, a character vector, and a matrix. Access each element of the list using <code>$</code> and <code>[[ ]]</code>. Modify one of the elements in the list (e.g., change a value in the matrix). Add a new element to the list (e.g., a logical vector).</p>
</section>
<section id="data-frames" class="level4">
<h4 class="anchored" data-anchor-id="data-frames">Data Frames</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a data frame with 3 columns 10 rows</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">gene =</span> <span class="fu">c</span>(<span class="st">"G1"</span>,<span class="st">"G2"</span>,<span class="st">"G3"</span>,<span class="st">"G4"</span>,<span class="st">"G5"</span>,<span class="st">"G6"</span>,<span class="st">"G7"</span>,<span class="st">"G8"</span>,<span class="st">"G9"</span>,<span class="st">"G10"</span>),</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">exp =</span>  <span class="fu">c</span>(<span class="fl">1.2</span>, <span class="fl">3.4</span>, <span class="fl">5.6</span>, <span class="fl">7.8</span>, <span class="fl">9.0</span>, <span class="fl">2.3</span>, <span class="fl">4.5</span>, <span class="fl">6.7</span>, <span class="fl">8.9</span>, <span class="fl">10.1</span>),</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">group =</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>),</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(df)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   10 obs. of  3 variables:
 $ gene : chr  "G1" "G2" "G3" "G4" ...
 $ exp  : num  1.2 3.4 5.6 7.8 9 2.3 4.5 6.7 8.9 10.1
 $ group: chr  "A" "B" "A" "B" ...</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(df)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>     gene                exp            group          
 Length:10          Min.   : 1.200   Length:10         
 Class :character   1st Qu.: 3.675   Class :character  
 Mode  :character   Median : 6.150   Mode  :character  
                    Mean   : 5.950                     
                    3rd Qu.: 8.625                     
                    Max.   :10.100                     </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Accessing data frame elements</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Accessing data frame elements</span></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>gene        <span class="co"># Access the "gene" column</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "G1"  "G2"  "G3"  "G4"  "G5"  "G6"  "G7"  "G8"  "G9"  "G10"</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>df[<span class="dv">1</span>, ]       <span class="co"># Access the first row</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  gene exp group
1   G1 1.2     A</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>df[ , <span class="dv">2</span>]      <span class="co"># Access the second column</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  1.2  3.4  5.6  7.8  9.0  2.3  4.5  6.7  8.9 10.1</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>df[df<span class="sc">$</span>exp <span class="sc">&gt;</span> <span class="dv">5</span>, ] <span class="co"># Subset rows where exp is greater than 5</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   gene  exp group
3    G3  5.6     A
4    G4  7.8     B
5    G5  9.0     A
8    G8  6.7     B
9    G9  8.9     A
10  G10 10.1     B</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>df[df<span class="sc">$</span>group <span class="sc">==</span> <span class="st">"A"</span>, ] <span class="co"># Subset rows where group is "A"</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  gene exp group
1   G1 1.2     A
3   G3 5.6     A
5   G5 9.0     A
7   G7 4.5     A
9   G9 8.9     A</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>df[df<span class="sc">$</span>exp <span class="sc">&gt;</span> <span class="dv">5</span> <span class="sc">&amp;</span> df<span class="sc">$</span>group <span class="sc">==</span> <span class="st">"A"</span>, ] <span class="co"># Subset rows where exp is greater than 5 and group is "A"</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  gene exp group
3   G3 5.6     A
5   G5 9.0     A
9   G9 8.9     A</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Adding a new column</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>Pass <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(df<span class="sc">$</span>exp <span class="sc">&gt;</span> <span class="dv">5</span>, <span class="st">"Pass"</span>, <span class="st">"Fail"</span>)</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>df</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   gene  exp group Pass
1    G1  1.2     A Fail
2    G2  3.4     B Fail
3    G3  5.6     A Pass
4    G4  7.8     B Pass
5    G5  9.0     A Pass
6    G6  2.3     B Fail
7    G7  4.5     A Fail
8    G8  6.7     B Pass
9    G9  8.9     A Pass
10  G10 10.1     B Pass</code></pre>
</div>
</div>
<p>Explanation Data Frames: A data frame is a table-like structure where each column can have a different data type (e.g., <code>numeric</code>, <code>character</code>). Accessing Columns: Use <code>$</code> to access a specific column by name. Subsetting Rows: Use logical conditions to filter rows based on specific criteria. Adding Columns: You can add new columns to a data frame using the <code>$</code> operator. In the example, a new column Pass is added based on the condition that exp is greater than 5. Exercise: Create a data frame with at least 5 columns and 10 rows. Access specific columns and rows using <code>$</code> and <code>[ ].</code> Subset the data frame based on specific conditions (e.g., values greater than a certain threshold).</p>
</section>
<section id="factors" class="level4">
<h4 class="anchored" data-anchor-id="factors">Factors</h4>
<p>Factors are used to represent categorical data in R. They are important for statistical modeling and data analysis. They can be ordered or unordered and are useful for representing categorical variables in data frames.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a factor variable</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">factor</span>(df<span class="sc">$</span>group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>), <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Group A"</span>, <span class="st">"Group B"</span>))</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>group</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] Group A Group B Group A Group B Group A Group B Group A Group B Group A
[10] Group B
Levels: Group A Group B</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the structure of the data frame</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(df)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   10 obs. of  4 variables:
 $ gene : chr  "G1" "G2" "G3" "G4" ...
 $ exp  : num  1.2 3.4 5.6 7.8 9 2.3 4.5 6.7 8.9 10.1
 $ group: Factor w/ 2 levels "Group A","Group B": 1 2 1 2 1 2 1 2 1 2
 $ Pass : chr  "Fail" "Fail" "Pass" "Pass" ...</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert a character vector to a factor</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">factor</span>(df<span class="sc">$</span>gene)</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>gene</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] G1  G2  G3  G4  G5  G6  G7  G8  G9  G10
Levels: G1 G10 G2 G3 G4 G5 G6 G7 G8 G9</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the levels of the factor</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(df<span class="sc">$</span>gene)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "G1"  "G10" "G2"  "G3"  "G4"  "G5"  "G6"  "G7"  "G8"  "G9" </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert a factor back to a character vector</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">as.character</span>(df<span class="sc">$</span>gene)</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>gene</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "G1"  "G2"  "G3"  "G4"  "G5"  "G6"  "G7"  "G8"  "G9"  "G10"</code></pre>
</div>
</div>
<p>Explanation Factors: Factors are used to represent categorical data in R. They are important for statistical modeling and data analysis. Creating Factors: Use the <code>factor()</code> function to create a factor variable. You can specify the levels and labels for better readability. Converting Factors: You can convert a factor back to a character vector using <code>as.character()</code>. Exercise: Create a factor variable for the group column in the data frame. Check the levels of the factor and convert it back to a character vector.</p>
</section>
</section>
<section id="wrangling-with-dplyr" class="level2">
<h2 class="anchored" data-anchor-id="wrangling-with-dplyr">5 <span id="dplyr">Wrangling with dplyr</span></h2>
<p>Explanation of dplyr Verbs The dplyr package provides a set of functions (verbs) for data manipulation. These functions are intuitive and work seamlessly with data frames and tibbles.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
</div>
<table class="caption-top table">
<colgroup>
<col style="width: 29%">
<col style="width: 32%">
<col style="width: 38%">
</colgroup>
<thead>
<tr class="header">
<th>Verb</th>
<th>Description</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>select()</code></td>
<td>choose columns</td>
<td>select(df, gene, expr)</td>
</tr>
<tr class="even">
<td><code>filter()</code></td>
<td>choose rows</td>
<td>filter(df, expr &gt; 6)</td>
</tr>
<tr class="odd">
<td><code>mutate()</code></td>
<td>add / transform column</td>
<td>mutate(df, log2expr = log2(expr))</td>
</tr>
<tr class="even">
<td><code>arrange()</code></td>
<td>sort rows</td>
<td>arrange(df, desc(expr))</td>
</tr>
<tr class="odd">
<td><code>group_by()</code></td>
<td>split into groups</td>
<td>group_by(df, gene)</td>
</tr>
<tr class="even">
<td><code>summarise()</code></td>
<td>aggregate per group</td>
<td>summarise(df, mean(expr))</td>
</tr>
</tbody>
</table>
<section id="select-choose-columns" class="level4">
<h4 class="anchored" data-anchor-id="select-choose-columns">5.1 select(): Choose Columns</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Select specific columns</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>df_selected <span class="ot">&lt;-</span> <span class="fu">select</span>(df, gene, exp)</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>df_selected</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   gene  exp
1    G1  1.2
2    G2  3.4
3    G3  5.6
4    G4  7.8
5    G5  9.0
6    G6  2.3
7    G7  4.5
8    G8  6.7
9    G9  8.9
10  G10 10.1</code></pre>
</div>
</div>
<p>Description: The <code>select()</code> function is used to choose specific columns from a data frame. Example: In the example above, only the gene and exp columns are selected from the data frame.</p>
</section>
<section id="filter-choose-rows" class="level4">
<h4 class="anchored" data-anchor-id="filter-choose-rows">5.2 filter(): Choose Rows</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter rows based on a condition</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>df_filtered <span class="ot">&lt;-</span> <span class="fu">filter</span>(df, exp <span class="sc">&gt;</span> <span class="dv">5</span>)</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>df_filtered</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  gene  exp   group Pass
1   G3  5.6 Group A Pass
2   G4  7.8 Group B Pass
3   G5  9.0 Group A Pass
4   G8  6.7 Group B Pass
5   G9  8.9 Group A Pass
6  G10 10.1 Group B Pass</code></pre>
</div>
</div>
<p>Description: The <code>filter()</code> function is used to subset rows based on logical conditions. Example: In the example above, only rows where exp is greater than 5 are selected. Exercise: Use <code>filter()</code> to extract rows where gene is ‚ÄúG1‚Äù. Try filtering rows based on multiple conditions (e.g., exp &gt; 5 and group == ‚ÄúA‚Äù).</p>
</section>
<section id="mutate-add-or-transform-columns" class="level4">
<h4 class="anchored" data-anchor-id="mutate-add-or-transform-columns">5.3 mutate(): Add or Transform Columns</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add a new column</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>df_mutated <span class="ot">&lt;-</span> <span class="fu">mutate</span>(df, <span class="at">Pass =</span> <span class="fu">ifelse</span>(exp <span class="sc">&gt;</span> <span class="dv">5</span>, <span class="st">"Pass"</span>, <span class="st">"Fail"</span>))</span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>df_mutated</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   gene  exp   group Pass
1    G1  1.2 Group A Fail
2    G2  3.4 Group B Fail
3    G3  5.6 Group A Pass
4    G4  7.8 Group B Pass
5    G5  9.0 Group A Pass
6    G6  2.3 Group B Fail
7    G7  4.5 Group A Fail
8    G8  6.7 Group B Pass
9    G9  8.9 Group A Pass
10  G10 10.1 Group B Pass</code></pre>
</div>
</div>
<p>Description: The <code>mutate()</code> function is used to add new columns or modify existing ones. Example: In the example above, a new column Pass is added based on the condition that exp is greater than 5. Exercise: Add a new column that calculates the square of the exp column. Try using <code>mutate()</code> to create a new column that categorizes exp into ‚ÄúLow‚Äù (&lt;= 5) and ‚ÄúHigh‚Äù (&gt; 5).</p>
</section>
<section id="arrange-sort-rows" class="level4">
<h4 class="anchored" data-anchor-id="arrange-sort-rows">5.4 arrange(): Sort Rows</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Sort rows by Score in descending order</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>df_arranged <span class="ot">&lt;-</span> <span class="fu">arrange</span>(df, <span class="fu">desc</span>(exp))</span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>df_arranged</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   gene  exp   group Pass
1   G10 10.1 Group B Pass
2    G5  9.0 Group A Pass
3    G9  8.9 Group A Pass
4    G4  7.8 Group B Pass
5    G8  6.7 Group B Pass
6    G3  5.6 Group A Pass
7    G7  4.5 Group A Fail
8    G2  3.4 Group B Fail
9    G6  2.3 Group B Fail
10   G1  1.2 Group A Fail</code></pre>
</div>
</div>
<p>Description: The <code>arrange()</code> function is used to sort rows based on one or more columns. Example: In the example above, rows are sorted by exp in descending order. Exercise: Sort rows by gene in ascending order. Try sorting by multiple columns (e.g., first by group and then by exp).</p>
</section>
<section id="group_by-and-summarise-group-and-aggregate" class="level4">
<h4 class="anchored" data-anchor-id="group_by-and-summarise-group-and-aggregate">5.5 group_by() and summarise(): Group and Aggregate</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>df <span class="sc">%&gt;%</span></span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group) <span class="sc">%&gt;%</span></span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_expr =</span> <span class="fu">mean</span>(exp),</span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">count     =</span> <span class="fu">n</span>())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 √ó 3
  group   mean_expr count
  &lt;fct&gt;       &lt;dbl&gt; &lt;int&gt;
1 Group A      5.84     5
2 Group B      6.06     5</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Group by Pass/Fail and calculate average expression</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>df_grouped <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Pass =</span> <span class="fu">ifelse</span>(exp <span class="sc">&gt;</span> <span class="dv">5</span>, <span class="st">"Pass"</span>, <span class="st">"Fail"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Pass) <span class="sc">%&gt;%</span></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_exp =</span> <span class="fu">mean</span>(exp))</span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>df_grouped</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 √ó 2
  Pass  mean_exp
  &lt;chr&gt;    &lt;dbl&gt;
1 Fail      2.85
2 Pass      8.02</code></pre>
</div>
</div>
<p>Description: The <code>group_by()</code> function is used to split the data into groups based on one or more columns. The <code>summarise()</code> function is then used to calculate summary statistics for each group. Example: In the example above, the data is grouped by gene, and the mean expression and count of rows are calculated for each group. Exercise: Group the data by group and calculate the mean expression for each group.</p>
</section>
</section>
<section id="reshaping-with-tidyr" class="level2">
<h2 class="anchored" data-anchor-id="reshaping-with-tidyr">6 <span id="tidyr">Reshaping with tidyr</span></h2>
<p>The tidyr package is used for reshaping and tidying data. It provides functions like <code>pivot_longer()</code> and <code>pivot_wider()</code> to transform data between wide and long formats. #### 6.1 pivot_longer(): Wide to Long Format</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example: Convert wide data to long format</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>wide <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">id =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="at">A =</span> <span class="fu">c</span>(<span class="dv">5</span>,<span class="dv">7</span>), <span class="at">B =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>))</span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>wide </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  id A B
1  1 5 2
2  2 7 3</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>long <span class="ot">&lt;-</span> wide <span class="sc">%&gt;%</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> A<span class="sc">:</span>B,</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to  =</span> <span class="st">"marker"</span>,</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"value"</span>)</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>long </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 √ó 3
     id marker value
  &lt;int&gt; &lt;chr&gt;  &lt;dbl&gt;
1     1 A          5
2     1 B          2
3     2 A          7
4     2 B          3</code></pre>
</div>
</div>
<p>Description: The <code>pivot_longer()</code> function converts wide-format data into long-format data by gathering multiple columns into key-value pairs. Example: In the example above, columns A and B are gathered into a single column marker, with their values stored in the value column. Exercise: Creat a wide data frame with 3 rows and 4 columns (e.g., id, X, Y, Z). Use <code>pivot_longer()</code> to convert it into long format. Rename the new columns to variable and measurement.</p>
<section id="pivot_wider-long-to-wide-format" class="level4">
<h4 class="anchored" data-anchor-id="pivot_wider-long-to-wide-format">6.2 pivot_wider(): Long to Wide Format</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example: Convert long data back to wide format</span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>wide_again <span class="ot">&lt;-</span> long <span class="sc">%&gt;%</span></span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> marker,</span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_from =</span> value)</span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>wide_again</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 √ó 3
     id     A     B
  &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt;
1     1     5     2
2     2     7     3</code></pre>
</div>
</div>
<p>Description: The <code>pivot_wider()</code> function converts long-format data into wide-format data by spreading key-value pairs into multiple columns. Example: In the example above, the marker column is spread into separate columns A and B. Exercise: Take the long-format data from the previous exercise and convert it back to wide format. Try using <code>pivot_wider()</code> with a dataset that has multiple grouping variables.</p>
</section>
</section>
<section id="visualising-with-ggplot2" class="level2">
<h2 class="anchored" data-anchor-id="visualising-with-ggplot2">7 <span id="ggplot">Visualising with ggplot2</span></h2>
<p>The <code>ggplot2</code> package is a powerful tool for creating visualizations in R. It uses a layered grammar of graphics to build plots step by step.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="histogram-density-plot" class="level4">
<h4 class="anchored" data-anchor-id="histogram-density-plot">7.1 Histogram + Density Plot</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(exp)) <span class="sc">+</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">1</span>, <span class="at">fill =</span> <span class="st">"#3182bd"</span>, <span class="at">colour =</span> <span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> .<span class="dv">3</span>, <span class="at">fill =</span> <span class="st">"orange"</span>) <span class="sc">+</span></span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Expression distribution"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Description: - <code>geom_histogram()</code>: Creates a histogram to visualize the distribution of expression values. - <code>geom_density():</code> Adds a density plot to show the distribution shape. - <code>binwidth</code>: Controls the width of the histogram bins. - <code>alpha</code>: Controls the transparency of the density plot. Exercise: - Create a histogram of the Age column with a binwidth of 5. - Add a density plot to the histogram. - Experiment with different fill colors and transparency levels.</p>
</section>
<section id="scatter-plot" class="level4">
<h4 class="anchored" data-anchor-id="scatter-plot">7.2 Scatter Plot</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a scatter plot</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(exp, <span class="fu">log2</span>(exp))) <span class="sc">+</span></span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">se =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb106-5"><a href="#cb106-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Scatter Plot of Expression vs log2(Expression)"</span>,</span>
<span id="cb106-6"><a href="#cb106-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Expression"</span>,</span>
<span id="cb106-7"><a href="#cb106-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"log2(Expression)"</span>) <span class="sc">+</span></span>
<span id="cb106-8"><a href="#cb106-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Description: <code>geom_point()</code>: Creates a scatter plot of expression vs log2(expression). <code>geom_smooth(method = "lm", se = FALSE)</code>: Adds a linear regression line without confidence intervals. <code>labs()</code>: Adds titles and labels to the axes. <code>theme_minimal()</code>: Applies a clean, minimal theme.</p>
</section>
<section id="bar-plot" class="level4">
<h4 class="anchored" data-anchor-id="bar-plot">7.3 Bar Plot</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a bar plot</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df , <span class="fu">aes</span>(<span class="at">x =</span> exp, <span class="at">fill =</span> Pass)) <span class="sc">+</span></span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_light</span>() <span class="sc">+</span></span>
<span id="cb108-5"><a href="#cb108-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Count of Pass and Fail"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Description: <code>geom_bar()</code>: Creates a bar plot. <code>aes(fill = Pass)</code>: Fills the bars with colors based on the Pass column. <code>theme_light()</code>: Applies a light theme. Exercise: Create a bar plot showing the count of Pass and Fail. Experiment with different themes (e.g., theme_classic(), theme_dark()).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> Pass, <span class="at">fill =</span> Pass)) <span class="sc">+</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_light</span>() <span class="sc">+</span></span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Count of Pass and Fail"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="full-workflow-on-biomed_data.csv" class="level2">
<h2 class="anchored" data-anchor-id="full-workflow-on-biomed_data.csv">8 <span id="workflow">üåü Full Workflow on biomed_data.csv</span></h2>
<section id="load-inspect" class="level4">
<h4 class="anchored" data-anchor-id="load-inspect">8.1 Load &amp; Inspect</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>biomed <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"biomed_data.csv"</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(biomed)      <span class="co"># columns &amp; types</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 200
Columns: 14
$ patient_id &lt;chr&gt; "P1", "P2", "P3", "P4", "P5", "P6", "P7", "P8", "P9", "P10"‚Ä¶
$ sex        &lt;chr&gt; "Female", "Female", "Female", "Female", "Male", "Female", "‚Ä¶
$ age        &lt;dbl&gt; 64, 27, 32, 43, 28, 44, 43, 31, 56, 41, 49, 59, 26, 61, 71,‚Ä¶
$ group      &lt;chr&gt; "Treated", "Treated", "Treated", "Control", "Control", "Con‚Ä¶
$ region     &lt;chr&gt; "North", "North", "North", "South", "South", "North", "Sout‚Ä¶
$ dose       &lt;chr&gt; "Low", "High", "High", "Low", "Medium", "High", "High", "Me‚Ä¶
$ marker_A   &lt;dbl&gt; 5.64, 3.61, 3.87, 3.86, 4.31, 5.17, 5.59, 4.18, 2.14, 5.95,‚Ä¶
$ marker_B   &lt;dbl&gt; 14.07, 11.44, 11.98, 13.23, 13.29, 10.17, 14.07, 8.11, 7.78‚Ä¶
$ marker_C   &lt;dbl&gt; 49.52, 46.27, 51.05, 45.76, 53.96, 50.15, 53.02, 45.03, 48.‚Ä¶
$ marker_D   &lt;dbl&gt; 1.23, 1.81, 1.91, 1.12, 0.99, 1.24, 1.54, 1.31, 0.79, 1.04,‚Ä¶
$ marker_E   &lt;dbl&gt; 11.38, 11.62, 10.65, 10.73, 9.75, 8.69, 10.27, 10.29, 9.00,‚Ä¶
$ expression &lt;dbl&gt; 8.66, 12.54, 11.99, 11.27, 8.67, 10.80, 8.90, 8.97, 6.92, 1‚Ä¶
$ heart_rate &lt;dbl&gt; 69, 88, 84, 64, 91, 100, 64, 85, 92, 91, 99, 99, 94, 94, 63‚Ä¶
$ RBC_count  &lt;dbl&gt; 4.47, 4.38, 3.99, 4.02, 5.32, 4.45, 4.73, 4.74, 4.95, 4.41,‚Ä¶</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(biomed)      <span class="co"># summary stats &amp; NAs</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  patient_id            sex                 age           group          
 Length:200         Length:200         Min.   :25.00   Length:200        
 Class :character   Class :character   1st Qu.:38.00   Class :character  
 Mode  :character   Mode  :character   Median :49.00   Mode  :character  
                                       Mean   :50.09                     
                                       3rd Qu.:61.25                     
                                       Max.   :75.00                     
                                                                         
    region              dose              marker_A        marker_B    
 Length:200         Length:200         Min.   :2.140   Min.   : 6.72  
 Class :character   Class :character   1st Qu.:4.300   1st Qu.:11.10  
 Mode  :character   Mode  :character   Median :4.900   Median :12.58  
                                       Mean   :4.896   Mean   :12.43  
                                       3rd Qu.:5.525   3rd Qu.:13.89  
                                       Max.   :7.350   Max.   :18.09  
                                       NA's   :5                      
    marker_C        marker_D         marker_E        expression    
 Min.   :35.35   Min.   :0.1400   Min.   : 5.850   Min.   : 4.000  
 1st Qu.:46.63   1st Qu.:0.8875   1st Qu.: 9.265   1st Qu.: 7.280  
 Median :49.93   Median :1.1200   Median :10.325   Median : 8.680  
 Mean   :49.89   Mean   :1.1838   Mean   :10.183   Mean   : 8.876  
 3rd Qu.:52.98   3rd Qu.:1.3550   3rd Qu.:11.110   3rd Qu.:10.340  
 Max.   :62.26   Max.   :2.9400   Max.   :15.040   Max.   :16.640  
 NA's   :5                                         NA's   :5       
   heart_rate       RBC_count    
 Min.   : 57.00   Min.   :3.550  
 1st Qu.: 67.00   1st Qu.:4.265  
 Median : 77.00   Median :4.550  
 Mean   : 78.89   Mean   :4.554  
 3rd Qu.: 91.00   3rd Qu.:4.860  
 Max.   :103.00   Max.   :5.490  
                  NA's   :5      </code></pre>
</div>
</div>
<p>Explanation - <code>read_csv()</code>: Reads a CSV file into a data frame. - <code>glimpse()</code>: Provides a quick overview of the data frame, including column names and types. - <code>summary()</code>: Provides summary statistics for each column, including mean, median, and number of missing values (NAs). - show_col_types = FALSE: Suppresses the display of column types in the output. - Exercise: - Load the <code>biomed_data.csv</code> file and inspect its structure. - Check for missing values in the dataset. - Summarize the data to understand its distribution and key statistics. - Identify any potential outliers or anomalies in the data.</p>
</section>
<section id="key-dplyr-demonstrations" class="level4">
<h4 class="anchored" data-anchor-id="key-dplyr-demonstrations">8.2 Key dplyr Demonstrations</h4>
<ol type="a">
<li>Select &amp; Filter</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep only patient_id, age, group, expression</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>biomed <span class="sc">%&gt;%</span></span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(patient_id, age, group, expression) <span class="sc">%&gt;%</span></span>
<span id="cb114-4"><a href="#cb114-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">5</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 √ó 4
  patient_id   age group   expression
  &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt;
1 P1            64 Treated       8.66
2 P2            27 Treated      12.5 
3 P3            32 Treated      12.0 
4 P4            43 Control      11.3 
5 P5            28 Control       8.67</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep only patient_id, age, group, expression</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>biomed_filtered <span class="ot">&lt;-</span> biomed <span class="sc">%&gt;%</span></span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(patient_id, age, group, expression) <span class="sc">%&gt;%</span></span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">30</span> <span class="sc">&amp;</span> group <span class="sc">==</span> <span class="st">"Treated"</span>)</span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>biomed_filtered</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 85 √ó 4
   patient_id   age group   expression
   &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt;
 1 P1            64 Treated       8.66
 2 P3            32 Treated      12.0 
 3 P8            31 Treated       8.97
 4 P11           49 Treated       7.88
 5 P15           71 Treated       9.45
 6 P16           71 Treated       6.17
 7 P17           54 Treated      12.7 
 8 P18           34 Treated      11.2 
 9 P20           73 Treated       7.29
10 P21           59 Treated       9.22
# ‚Ñπ 75 more rows</code></pre>
</div>
</div>
<p>Description: - <code>select()</code>: Keeps only the specified columns (patient_id, age, group, expression). - <code>filter()</code>: Filters rows where age &gt; 30 and group is ‚ÄúTreated‚Äù. Exercise: Select only the patient_id, group, and expression columns. Filter rows where expression is greater than 50 and group is ‚ÄúControl‚Äù.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Patients older than 60 in Treated group</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>biomed <span class="sc">%&gt;%</span></span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">60</span>, group <span class="sc">==</span> <span class="st">"Treated"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(patient_id, age, group)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 29 √ó 3
   patient_id   age group  
   &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;  
 1 P1            64 Treated
 2 P15           71 Treated
 3 P16           71 Treated
 4 P20           73 Treated
 5 P27           61 Treated
 6 P30           71 Treated
 7 P40           72 Treated
 8 P61           66 Treated
 9 P62           61 Treated
10 P64           66 Treated
# ‚Ñπ 19 more rows</code></pre>
</div>
</div>
<p>Description: - <code>filter()</code>: Filters rows where age is greater than 60 and group is ‚ÄúTreated‚Äù. - <code>select()</code>: Keeps only the patient_id, age, and group columns. Exercise: - Filter patients with expression &gt; 10 and group ‚ÄúControl‚Äù. - Select only the patient_id, age, and expression columns.</p>
<ol start="2" type="a">
<li>Mutate &amp; Arrange</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add a new column for age category and sort by expression</span></span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>biomed_mutated <span class="ot">&lt;-</span> biomed <span class="sc">%&gt;%</span></span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">age_category =</span> <span class="fu">ifelse</span>(age <span class="sc">&gt;</span> <span class="dv">50</span>, <span class="st">"Senior"</span>, <span class="st">"Adult"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb120-4"><a href="#cb120-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(expression))</span>
<span id="cb120-5"><a href="#cb120-5" aria-hidden="true" tabindex="-1"></a>biomed_mutated</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 200 √ó 15
   patient_id sex     age group region dose  marker_A marker_B marker_C marker_D
   &lt;chr&gt;      &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
 1 P54        Fema‚Ä¶    59 Trea‚Ä¶ North  High      3.31    12.8      NA       2.15
 2 P127       Fema‚Ä¶    29 Trea‚Ä¶ East   High      5.19    15.1      51.3     2.05
 3 P30        Fema‚Ä¶    71 Trea‚Ä¶ South  High      4.57     9.95     43.8     1.28
 4 P168       Fema‚Ä¶    70 Trea‚Ä¶ South  Low       2.34    16.4      54.5     0.14
 5 P91        Male     37 Trea‚Ä¶ North  High      6.26    13.4      53.4     1.82
 6 P140       Fema‚Ä¶    48 Trea‚Ä¶ North  High      5.25    14.1      44.1     2.64
 7 P10        Fema‚Ä¶    41 Cont‚Ä¶ North  High      5.95    12.5      51.7     1.04
 8 P165       Fema‚Ä¶    48 Trea‚Ä¶ North  High      5.14    12.4      49.3     1.96
 9 P79        Fema‚Ä¶    68 Trea‚Ä¶ East   High      4.62    12.5      53.6     1.99
10 P102       Male     62 Trea‚Ä¶ West   High      4.35    11.0      48.9     2.94
# ‚Ñπ 190 more rows
# ‚Ñπ 5 more variables: marker_E &lt;dbl&gt;, expression &lt;dbl&gt;, heart_rate &lt;dbl&gt;,
#   RBC_count &lt;dbl&gt;, age_category &lt;chr&gt;</code></pre>
</div>
</div>
<p>Description: - <code>mutate()</code>: Adds a new column (age_category) that categorizes age into ‚ÄúSenior‚Äù (&gt;50) and ‚ÄúAdult‚Äù (&lt;=50). - <code>arrange()</code>: Sorts the data frame by expression in descending order.</p>
<p>Exercise: Create a new column that categorizes expression into ‚ÄúHigh‚Äù (&gt;7) and ‚ÄúLow‚Äù (&lt;=7). Sort the data by age in ascending order.</p>
<ol start="3" type="a">
<li>Group &amp; Summarise</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Mean expression &amp; count per group &amp; dose</span></span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>biomed <span class="sc">%&gt;%</span></span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group, dose) <span class="sc">%&gt;%</span></span>
<span id="cb122-4"><a href="#cb122-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_expr =</span> <span class="fu">mean</span>(expression, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb122-5"><a href="#cb122-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">sd_expr   =</span> <span class="fu">sd</span>(expression, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb122-6"><a href="#cb122-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">n         =</span> <span class="fu">n</span>())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has regrouped the output.
‚Ñπ Summaries were computed grouped by group and dose.
‚Ñπ Output is grouped by group.
‚Ñπ Use `summarise(.groups = "drop_last")` to silence this message.
‚Ñπ Use `summarise(.by = c(group, dose))` for per-operation grouping
  (`?dplyr::dplyr_by`) instead.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 √ó 5
# Groups:   group [2]
  group   dose   mean_expr sd_expr     n
  &lt;chr&gt;   &lt;chr&gt;      &lt;dbl&gt;   &lt;dbl&gt; &lt;int&gt;
1 Control High        8.42    2.01    37
2 Control Low         8.63    1.79    34
3 Control Medium      7.89    1.88    34
4 Treated High       11.8     1.84    34
5 Treated Low         8.41    1.83    32
6 Treated Medium      7.90    1.77    29</code></pre>
</div>
</div>
<p>Description: - <code>group_by()</code>: Groups the data by group and dose. - <code>summarise()</code>: Calculates the mean expression, standard deviation, and count of rows for each group and dose. Exercise: - Group the data by group and dose, and calculate the mean expression and standard deviation for each group.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Group by group and calculate mean expression</span></span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a>biomed_summary <span class="ot">&lt;-</span> biomed <span class="sc">%&gt;%</span></span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group) <span class="sc">%&gt;%</span></span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_expression =</span> <span class="fu">mean</span>(expression, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb125-5"><a href="#cb125-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">count =</span> <span class="fu">n</span>())</span>
<span id="cb125-6"><a href="#cb125-6" aria-hidden="true" tabindex="-1"></a>biomed_summary</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 √ó 3
  group   mean_expression count
  &lt;chr&gt;             &lt;dbl&gt; &lt;int&gt;
1 Control            8.31   105
2 Treated            9.50    95</code></pre>
</div>
</div>
<p>Description: <code>group_by()</code>: Groups the data by the group column. <code>summarise()</code>: Calculates the mean expression and the count of rows for each group. Exercise: Group the data by age_category and calculate the median expression for each category. Add a column to count the number of patients in each group.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Frequency table of region</span></span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a>biomed <span class="sc">%&gt;%</span></span>
<span id="cb127-3"><a href="#cb127-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(region) <span class="sc">%&gt;%</span></span>
<span id="cb127-4"><a href="#cb127-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(n))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 √ó 2
  region     n
  &lt;chr&gt;  &lt;int&gt;
1 East      59
2 North     51
3 South     50
4 West      40</code></pre>
</div>
</div>
<ol start="4" type="a">
<li>Full Pipeline Example</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine all steps into a single pipeline    &lt;!-- filter(age &gt; 30 &amp; group == "Treatment") %&gt;% --&gt;</span></span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>biomed_pipeline <span class="ot">&lt;-</span> biomed <span class="sc">%&gt;%</span></span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(patient_id, age, group, expression) <span class="sc">%&gt;%</span></span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb129-5"><a href="#cb129-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">age_category =</span> <span class="fu">ifelse</span>(age <span class="sc">&gt;</span> <span class="dv">50</span>, <span class="st">"Senior"</span>, <span class="st">"Young"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb129-6"><a href="#cb129-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(age_category) <span class="sc">%&gt;%</span></span>
<span id="cb129-7"><a href="#cb129-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_expression =</span> <span class="fu">mean</span>(expression, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb129-8"><a href="#cb129-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">count =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb129-9"><a href="#cb129-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(mean_expression))</span>
<span id="cb129-10"><a href="#cb129-10" aria-hidden="true" tabindex="-1"></a>biomed_pipeline</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 √ó 3
  age_category mean_expression count
  &lt;chr&gt;                  &lt;dbl&gt; &lt;int&gt;
1 Young                   9.37   101
2 Senior                  8.36    99</code></pre>
</div>
</div>
<p>Description: This pipeline combines <code>select()</code>, <code>filter()</code>, <code>mutate()</code>, <code>group_by()</code>, <code>summarise()</code>, and <code>arrange()</code> into a single workflow. Exercise: Modify the pipeline to include only patients with expression &gt; 50. Add a step to calculate the standard deviation of expression for each age_category.</p>
</section>
<section id="extensive-ggplot2-gallery" class="level4">
<h4 class="anchored" data-anchor-id="extensive-ggplot2-gallery">8.3 Extensive ggplot2 Gallery</h4>
<p>Histogram of Age</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(biomed, <span class="fu">aes</span>(age)) <span class="sc">+</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">5</span>, <span class="at">fill =</span> <span class="st">"#3182bd"</span>, <span class="at">colour =</span> <span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Age distribution (binwidth=5)"</span>, <span class="at">x =</span> <span class="st">"Age"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Explanation - <code>geom_histogram()</code>: Creates a histogram to visualize the distribution of age values. - <code>binwidth</code>: Controls the width of the histogram bins. - <code>labs()</code>: Adds titles and labels to the axes. - Exercise: - Create a histogram of expression with a binwidth of 10. - Add a density plot to the histogram. - Experiment with different fill colors and transparency levels.</p>
<p>Density of Expression by Group</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(biomed, <span class="fu">aes</span>(expression, <span class="at">fill =</span> group)) <span class="sc">+</span></span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> .<span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Expression density by group"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing non-finite outside the scale range
(`stat_density()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Explanation - <code>geom_density()</code>: Creates a density plot to visualize the distribution of expression values by group. - <code>alpha</code>: Controls the transparency of the density plot. - <code>scale_fill_viridis_d()</code>: Applies a color palette for better visibility. - Exercise: - Create a density plot of heart_rate by region. - Experiment with different fill colors and transparency levels.</p>
<ol type="a">
<li>Line Plot</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Line plot showing trends over time</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(biomed, <span class="fu">aes</span>(<span class="at">x =</span> age, <span class="at">y =</span> expression, <span class="at">group =</span> group, <span class="at">colour =</span> group)) <span class="sc">+</span></span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb134-5"><a href="#cb134-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Expression over Age by Group"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>xplanation - <code>geom_line()</code>: Creates a line plot to visualize trends over time (or age). - <code>geom_point()</code>: Adds points to the line plot for better visibility. - <code>group</code>: Groups the data by the group variable. - <code>colour</code>: Colors the lines and points based on the group variable. - Exercise: - Create a line plot of heart_rate over age by region.</p>
<ol start="2" type="a">
<li>Boxplot: Expression by Dose &amp; Group</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(biomed, <span class="fu">aes</span>(dose, expression, <span class="at">fill =</span> group)) <span class="sc">+</span></span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">width =</span> .<span class="dv">8</span>)) <span class="sc">+</span></span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb136-4"><a href="#cb136-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Expression by dose &amp; group"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing non-finite outside the scale range
(`stat_boxplot()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Explanation - <code>geom_boxplot()</code>: Creates a boxplot to visualize the distribution of expression values by dose and group. - <code>position_dodge()</code>: Adjusts the position of the boxplots to avoid overlap. - <code>scale_fill_viridis_d()</code>: Applies a color palette for better visibility. Exercise: - Create a boxplot for marker_A by region. - Add jittered points to the boxplot to show individual data points.</p>
<ol start="3" type="a">
<li>Violin: Marker_D by Region</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(biomed, <span class="fu">aes</span>(region, marker_D, <span class="at">fill =</span> region)) <span class="sc">+</span></span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">trim =</span> <span class="cn">FALSE</span>, <span class="at">alpha =</span> .<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> .<span class="dv">1</span>, <span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb138-4"><a href="#cb138-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb138-5"><a href="#cb138-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Marker_D distribution by region"</span>) <span class="sc">+</span></span>
<span id="cb138-6"><a href="#cb138-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Explanation - <code>geom_violin()</code>: Creates a violin plot to visualize the distribution of marker_D values by region. - <code>geom_boxplot()</code>: Adds a boxplot inside the violin plot for better summary statistics. - trim = FALSE: Ensures the violin plot is not trimmed at the tails. - <code>theme(axis.text.x = element_text(angle = 45, hjust = 1))</code>: Rotates x-axis labels for better readability. Exercise: - Create a violin plot for marker_E by group. - Add a boxplot inside the violin plot to show summary statistics.</p>
<ol start="4" type="a">
<li>Scatter: marker_B vs marker_A, Faceted by Dose</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(biomed, <span class="fu">aes</span>(marker_A, marker_B, <span class="at">colour =</span> dose)) <span class="sc">+</span></span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> .<span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">se =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> dose) <span class="sc">+</span></span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"marker_B vs marker_A by dose"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing non-finite outside the scale range
(`stat_smooth()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Explanation - <code>geom_point()</code>: Creates a scatter plot of marker_B vs marker_A. - <code>geom_smooth(method = "lm", se = FALSE)</code>: Adds a linear regression line without confidence intervals. - <code>facet_wrap(~ dose)</code>: Creates separate panels for each dose level. - <code>scale_colour_viridis_d()</code>: Applies a color palette for better visibility. Exercise: - Create a scatter plot of heart_rate vs age, colored by group. - Add a linear regression line to the scatter plot.</p>
<ol start="5" type="a">
<li>Jittered Scatter: Heart Rate vs Age</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(biomed, <span class="fu">aes</span>(age, heart_rate, <span class="at">colour =</span> group)) <span class="sc">+</span></span>
<span id="cb143-2"><a href="#cb143-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">width =</span> .<span class="dv">5</span>, <span class="at">height =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb143-3"><a href="#cb143-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb143-4"><a href="#cb143-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Heart rate vs age (jittered)"</span>) <span class="sc">+</span></span>
<span id="cb143-5"><a href="#cb143-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Explanation - <code>geom_jitter()</code>: Creates a scatter plot with jitter to avoid overplotting. - <code>width and height</code>: Control the amount of jitter in the x and y directions. - <code>theme_minimal()</code>: Applies a clean, minimal theme. Exercise: - Create a jittered scatter plot of marker_C vs age, colored by region.</p>
<ol start="6" type="a">
<li>Facet Grid: marker_E by Group &amp; Dose</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(biomed, <span class="fu">aes</span>(marker_E, <span class="at">fill =</span> group)) <span class="sc">+</span></span>
<span id="cb144-2"><a href="#cb144-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">20</span>, <span class="at">alpha =</span> .<span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb144-3"><a href="#cb144-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(group <span class="sc">~</span> dose) <span class="sc">+</span></span>
<span id="cb144-4"><a href="#cb144-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb144-5"><a href="#cb144-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"marker_E distribution, group √ó dose"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Explanation - <code>facet_grid(group ~ dose)</code>: Creates a grid of plots based on group and dose. - <code>geom_histogram(bins = 20, alpha = .6)</code>: Creates histograms for marker_E with 20 bins. - <code>scale_fill_viridis_d()</code>: Applies a color palette for better visibility. Exercise: - Create a facet grid of marker_A by region and group.</p>
<ol start="7" type="a">
<li>Bar Chart: Count per Region &amp; Group</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(biomed, <span class="fu">aes</span>(region, <span class="at">fill =</span> group)) <span class="sc">+</span></span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">position =</span> <span class="st">"dodge"</span>) <span class="sc">+</span></span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb145-4"><a href="#cb145-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Patient counts by region &amp; group"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Explanation - <code>geom_bar(position = "dodge")</code>: Creates a bar chart showing counts of patients by region and group. - <code>scale_fill_viridis_d()</code>: Applies a color palette for better visibility. - <code>labs(title = "Patient counts by region &amp; group")</code>: Adds a title to the plot. Exercise: - Create a bar chart showing counts of patients by age category and group. h) Pairwise Scatterplot Matrix</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GGally)</span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a pairwise scatterplot matrix</span></span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggpairs</span>(biomed, <span class="at">columns =</span> <span class="fu">c</span>(<span class="st">"marker_A"</span>, <span class="st">"marker_B"</span>, <span class="st">"marker_C"</span>, <span class="st">"marker_D"</span>),</span>
<span id="cb146-4"><a href="#cb146-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">colour =</span> group, <span class="at">alpha =</span> <span class="fl">0.5</span>)) <span class="sc">+</span></span>
<span id="cb146-5"><a href="#cb146-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb146-6"><a href="#cb146-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Pairwise scatterplot matrix"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing non-finite outside the scale range
(`stat_density()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing missing values</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 10 rows containing missing values</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing missing values</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing missing values</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 10 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing non-finite outside the scale range
(`stat_density()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing missing values</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 5 rows containing missing values or values outside the scale range
(`geom_point()`).
Removed 5 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Explanation - <code>ggpairs()</code>: Creates a matrix of scatterplots for selected columns. - <code>aes(colour = group, alpha = 0.5)</code>: Colors the points based on the group and sets transparency. - <code>theme_minimal()</code>: Applies a clean, minimal theme. - <code>labs(title = "Pairwise scatterplot matrix")</code>: Adds a title to the plot. Exercise: - Create a pairwise scatterplot matrix for heart_rate, RBC_count, and expression. i) Correlation Matrix</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb158"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute correlation matrix</span></span>
<span id="cb158-2"><a href="#cb158-2" aria-hidden="true" tabindex="-1"></a>cor_matrix <span class="ot">&lt;-</span> <span class="fu">cor</span>(biomed[, <span class="fu">c</span>(<span class="st">"marker_A"</span>, <span class="st">"marker_B"</span>, <span class="st">"marker_C"</span>, <span class="st">"marker_D"</span>)], <span class="at">use =</span> <span class="st">"pairwise.complete.obs"</span>)</span>
<span id="cb158-3"><a href="#cb158-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to long form</span></span>
<span id="cb158-4"><a href="#cb158-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb158-5"><a href="#cb158-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert correlation matrix to long format without using rownames_to_column</span></span>
<span id="cb158-6"><a href="#cb158-6" aria-hidden="true" tabindex="-1"></a>Long <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">as.table</span>(cor_matrix))</span>
<span id="cb158-7"><a href="#cb158-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot heatmap</span></span>
<span id="cb158-8"><a href="#cb158-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(Long, <span class="fu">aes</span>(Var1, Var2, <span class="at">fill =</span> Freq)) <span class="sc">+</span></span>
<span id="cb158-9"><a href="#cb158-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>() <span class="sc">+</span></span>
<span id="cb158-10"><a href="#cb158-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"white"</span>, <span class="at">high =</span> <span class="st">"red"</span>,</span>
<span id="cb158-11"><a href="#cb158-11" aria-hidden="true" tabindex="-1"></a>                       <span class="at">midpoint =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb158-12"><a href="#cb158-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Correlation heatmap"</span>) <span class="sc">+</span></span>
<span id="cb158-13"><a href="#cb158-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>, <span class="at">hjust=</span><span class="dv">1</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Explanation - <code>cor()</code>: Computes the correlation matrix for selected columns. - <code>as.table()</code>: Converts the correlation matrix to a table format. - <code>ggplot()</code>: Creates a heatmap of the correlation matrix. - <code>geom_tile()</code>: Creates a heatmap of the correlation matrix. - <code>scale_fill_gradient2()</code>: Sets the color gradient for the heatmap. - <code>labs(title = "Correlation heatmap")</code>: Adds a title to the plot. - <code>theme(axis.text.x = element_text(angle=45, hjust=1))</code>: Rotates x-axis labels for better readability. Exercise: - Create a correlation matrix for heart_rate, RBC_count, and expression. - Plot the correlation matrix as a heatmap.</p>
<ol start="10" type="a">
<li>Correlation Heatmap of Numeric Variables</li>
</ol>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute correlation matrix</span></span>
<span id="cb159-2"><a href="#cb159-2" aria-hidden="true" tabindex="-1"></a>nums <span class="ot">&lt;-</span> biomed <span class="sc">%&gt;%</span></span>
<span id="cb159-3"><a href="#cb159-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(marker_A, marker_B, marker_C, marker_D,</span>
<span id="cb159-4"><a href="#cb159-4" aria-hidden="true" tabindex="-1"></a>         marker_E, expression, heart_rate, RBC_count) <span class="sc">%&gt;%</span></span>
<span id="cb159-5"><a href="#cb159-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cor</span>(<span class="at">use =</span> <span class="st">"pairwise.complete.obs"</span>)</span>
<span id="cb159-6"><a href="#cb159-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb159-7"><a href="#cb159-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to long form</span></span>
<span id="cb159-8"><a href="#cb159-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb159-9"><a href="#cb159-9" aria-hidden="true" tabindex="-1"></a><span class="co"># corr_long &lt;- as.data.frame(nums) %&gt;%</span></span>
<span id="cb159-10"><a href="#cb159-10" aria-hidden="true" tabindex="-1"></a><span class="co">#   rownames_to_column("var1") %&gt;%</span></span>
<span id="cb159-11"><a href="#cb159-11" aria-hidden="true" tabindex="-1"></a><span class="co">#   pivot_longer(-var1, names_to="var2", values_to="corr")</span></span>
<span id="cb159-12"><a href="#cb159-12" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb159-13"><a href="#cb159-13" aria-hidden="true" tabindex="-1"></a><span class="co"># # Plot heatmap</span></span>
<span id="cb159-14"><a href="#cb159-14" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot(corr_long, aes(var1, var2, fill = corr)) +</span></span>
<span id="cb159-15"><a href="#cb159-15" aria-hidden="true" tabindex="-1"></a><span class="co">#   geom_tile() +</span></span>
<span id="cb159-16"><a href="#cb159-16" aria-hidden="true" tabindex="-1"></a><span class="co">#   scale_fill_gradient2(low = "blue", mid = "white", high = "red",</span></span>
<span id="cb159-17"><a href="#cb159-17" aria-hidden="true" tabindex="-1"></a><span class="co">#                        midpoint = 0) +</span></span>
<span id="cb159-18"><a href="#cb159-18" aria-hidden="true" tabindex="-1"></a><span class="co">#   labs(title = "Correlation heatmap") +</span></span>
<span id="cb159-19"><a href="#cb159-19" aria-hidden="true" tabindex="-1"></a><span class="co">#   theme(axis.text.x = element_text(angle=45, hjust=1))</span></span>
<span id="cb159-20"><a href="#cb159-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb159-21"><a href="#cb159-21" aria-hidden="true" tabindex="-1"></a>Long <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">as.table</span>(nums))</span>
<span id="cb159-22"><a href="#cb159-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot heatmap</span></span>
<span id="cb159-23"><a href="#cb159-23" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(Long, <span class="fu">aes</span>(Var1, Var2, <span class="at">fill =</span> Freq)) <span class="sc">+</span></span>
<span id="cb159-24"><a href="#cb159-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>() <span class="sc">+</span></span>
<span id="cb159-25"><a href="#cb159-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"white"</span>, <span class="at">high =</span> <span class="st">"red"</span>,</span>
<span id="cb159-26"><a href="#cb159-26" aria-hidden="true" tabindex="-1"></a>                       <span class="at">midpoint =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb159-27"><a href="#cb159-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Correlation heatmap"</span>) <span class="sc">+</span></span>
<span id="cb159-28"><a href="#cb159-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>, <span class="at">hjust=</span><span class="dv">1</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="R_Programming_basics_practical_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Explanation - <code>cor()</code>: Computes the correlation matrix for selected columns. - <code>rownames_to_column()</code>: Converts row names to a column for easier plotting. - <code>pivot_longer()</code>: Reshapes the correlation matrix into long format. - <code>geom_tile()</code>: Creates a heatmap of the correlation matrix. - <code>scale_fill_gradient2()</code>: Sets the color gradient for the heatmap. - <code>labs(title = "Correlation heatmap")</code>: Adds a title to the plot. Exercise: - Create a correlation matrix for heart_rate, RBC_count, and expression. - Plot the correlation matrix as a heatmap.</p>
</section>
<section id="references-resources" class="level3">
<h3 class="anchored" data-anchor-id="references-resources">References &amp; Resources</h3>
<p>Below is a curated list of books, websites, cheat‚Äêsheets and tutorials to deepen your understanding of R, the tidyverse, and biomedical data analysis.</p>
<hr>
<section id="books" class="level4">
<h4 class="anchored" data-anchor-id="books">Books</h4>
<ul>
<li><p><strong>R for Data Science</strong><br>
Wickham, H. &amp; Grolemund, G. (2016). <em>R for Data Science</em>. O‚ÄôReilly Media.<br>
Online: <a href="https://r4ds.had.co.nz/" class="uri">https://r4ds.had.co.nz/</a> :contentReference<span data-index="0">oaicite:0</span></p></li>
<li><p><strong>ggplot2: Elegant Graphics for Data Analysis</strong><br>
Wickham, H. (2016). <em>ggplot2: Elegant Graphics for Data Analysis</em>. Springer.<br>
Online reference: <a href="https://ggplot2.tidyverse.org/" class="uri">https://ggplot2.tidyverse.org/</a> :contentReference<span data-index="1">oaicite:1</span></p></li>
<li><p><strong>Advanced R</strong><br>
Wickham, H. (2019). <em>Advanced R</em>. Chapman &amp; Hall/CRC.<br>
Online: <a href="https://adv-r.hadley.nz/" class="uri">https://adv-r.hadley.nz/</a> :contentReference<span data-index="2">oaicite:2</span></p></li>
</ul>
<hr>
</section>
<section id="official-documentation" class="level4">
<h4 class="anchored" data-anchor-id="official-documentation">Official Documentation</h4>
<ul>
<li><p><strong>tidyverse</strong> (‚Äúmeta‚Äêpackage‚Äù including dplyr, tidyr, ggplot2, readr, etc.)<br>
<a href="https://www.tidyverse.org/" class="uri">https://www.tidyverse.org/</a> :contentReference<span data-index="3">oaicite:3</span></p></li>
<li><p><strong>dplyr reference</strong><br>
<a href="https://dplyr.tidyverse.org/reference/" class="uri">https://dplyr.tidyverse.org/reference/</a> :contentReference<span data-index="4">oaicite:4</span></p></li>
<li><p><strong>tidyr reference</strong><br>
<a href="https://tidyr.tidyverse.org/reference/" class="uri">https://tidyr.tidyverse.org/reference/</a> :contentReference<span data-index="5">oaicite:5</span></p></li>
<li><p><strong>ggplot2 reference</strong><br>
<a href="https://ggplot2.tidyverse.org/reference/" class="uri">https://ggplot2.tidyverse.org/reference/</a> :contentReference<span data-index="6">oaicite:6</span></p></li>
<li><p><strong>readr reference</strong><br>
<a href="https://readr.tidyverse.org/reference/" class="uri">https://readr.tidyverse.org/reference/</a> :contentReference<span data-index="7">oaicite:7</span></p></li>
</ul>
<hr>
</section>
<section id="cheatsheets-quick-guides" class="level4">
<h4 class="anchored" data-anchor-id="cheatsheets-quick-guides">Cheat‚ÄêSheets &amp; Quick Guides</h4>
<ul>
<li><p><strong>RStudio Cheat‚ÄêSheets</strong><br>
Download PDF versions for dplyr, tidyr, ggplot2, and more:<br>
<a href="https://www.rstudio.com/resources/cheatsheets/" class="uri">https://www.rstudio.com/resources/cheatsheets/</a> :contentReference<span data-index="8">oaicite:8</span></p></li>
<li><p><strong>Tidyverse Style Guide</strong><br>
Conventions for writing tidyverse‚Äêstyle R code:<br>
<a href="https://style.tidyverse.org/" class="uri">https://style.tidyverse.org/</a> :contentReference<span data-index="9">oaicite:9</span></p></li>
</ul>
<hr>
</section>
</section>
</section>
<section id="online-tutorials-courses" class="level2">
<h2 class="anchored" data-anchor-id="online-tutorials-courses">Online Tutorials &amp; Courses</h2>
<ul>
<li><p><strong>Swirl</strong>: Interactive R lessons in the R console<br>
<a href="https://swirlstats.com/" class="uri">https://swirlstats.com/</a> :contentReference<span data-index="10">oaicite:10</span></p></li>
<li><p><strong>Coursera: R Programming</strong> (Johns Hopkins University)<br>
<a href="https://www.coursera.org/learn/r-programming" class="uri">https://www.coursera.org/learn/r-programming</a> :contentReference<span data-index="11">oaicite:11</span></p></li>
<li><p><strong>Datacamp: Introduction to R</strong><br>
<a href="https://www.datacamp.com/courses/free-introduction-to-r" class="uri">https://www.datacamp.com/courses/free-introduction-to-r</a> :contentReference<span data-index="12">oaicite:12</span></p></li>
</ul>
<hr>
<section id="biomedical-specific-resources" class="level4">
<h4 class="anchored" data-anchor-id="biomedical-specific-resources">Biomedical ‚Äê Specific Resources</h4>
<ul>
<li><p><strong>Bioconductor</strong>: R packages for bioinformatics<br>
<a href="https://www.bioconductor.org/" class="uri">https://www.bioconductor.org/</a> :contentReference<span data-index="13">oaicite:13</span></p></li>
<li><p><strong>Bioconductor Workflow Guides</strong><br>
E.g. RNA‚Äêseq, single‚Äêcell analyses:<br>
<a href="https://www.bioconductor.org/packages/release/BiocViews.html#___Workflow" class="uri">https://www.bioconductor.org/packages/release/BiocViews.html#___Workflow</a> :contentReference<span data-index="14">oaicite:14</span></p></li>
</ul>
<hr>
</section>
<section id="community-help" class="level4">
<h4 class="anchored" data-anchor-id="community-help">Community &amp; Help</h4>
<ul>
<li><p><strong>RStudio Community</strong><br>
<a href="https://community.rstudio.com/" class="uri">https://community.rstudio.com/</a> :contentReference<span data-index="15">oaicite:15</span></p></li>
<li><p><strong>Stack Overflow (R tag)</strong><br>
<a href="https://stackoverflow.com/questions/tagged/r" class="uri">https://stackoverflow.com/questions/tagged/r</a> :contentReference<span data-index="16">oaicite:16</span></p></li>
<li><p><strong>R‚ÄêBloggers</strong> (aggregated R tutorials and news)<br>
<a href="https://www.r-bloggers.com/" class="uri">https://www.r-bloggers.com/</a> :contentReference<span data-index="17">oaicite:17</span></p></li>
</ul>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "Óßã";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>